Add size, lag, and limit metrics

katcharov opened the pull request:
Uses a dynamic MBean to help manage the additional metrics. Some of the metrics have been renamed to better fit as prefixes for the new metrics. The MBean naming is now more consistent with Kafka:

Mongo: `com.mongodb.kafka.connect:type=sink-task-metrics,task=sink-task-0`
Kafka: `kafka.connect:type=sink-task-metrics,connector=MongoSinkConnector,task=0`

The "0" is taken from the current thread, which Kafka names (internally) according to the task number. The stats are logged at the debug level at the end of each task invocation.

Search for `.sample(` in the code changes to see where the numbers come from. The integration tests enumerate the metrics.

https://jira.mongodb.org/browse/KAFKA-64

katcharov commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/AdditiveMetric.java:
```diff
@@ -0,0 +1,60 @@
+      consumer.accept(
+          new MetricValue(name + "-over-" + limits[i] + unit, () -> bins[finalI].get()));
+    }
+    consumer.accept(new MetricValue(name + "-total-" + unit, () -> this.total.get()));
```
This produces:
```
"task-invocations",
"task-invocations-over-1ms",
"task-invocations-over-10ms",
"task-invocations-over-100ms",
"task-invocations-over-1000ms",
"task-invocations-over-10000ms",
"task-invocations-total-ms",
```
We could also add:
```
"task-invocations-max-ms"
"task-invocations-latest-ms"
```
(These would be added to all 11 "additive" metrics.)

jyemin reviewed (commented):
I didn't remember what a DynamicMBean was, and when I looked it up I found something to discuss.  Let's resolve that thread before proceeding.

jyemin commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MongoMBean.java:
```diff
@@ -0,0 +1,142 @@
+  @Override
+  public MBeanInfo getMBeanInfo() {
+    List<MBeanAttributeInfo> attrs = new ArrayList<>();
+    for (Metric metric : this.metrics) {
```
What's weird is that the DynamicMBean Javadoc doesn't state this requirement.

katcharov replied:
In the current code, the MBean never changes. Each metric is registered with the MBean on field initialization, before the MBean is registered (and therefore visible for an `getMBeanInfo` invocation). This registration pattern is similar to [what Kafka does](https://github.com/a0x8o/kafka/blob/master/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java#L385), and allows fields to be specified as follows:
```
  private Metric failedRecords = registerTotal("records-failed");
  private Metric latestOffsetMs = registerLatest("latest-offset-ms");
  private Metric taskInvocations = registerMs("task-invocations");
```
It is true that we might later decide to register a new metric after the MBean has been constructed, and inadvertently violate this constraint, but this seems improbable, and difficult to prevent in code regardless.

Our tests already assert that each MBean produces a complete set of all metric attributes, after the test completes. I can add a test that confirms that all metric attributes available immediately after the MBean constructor is invoked match that complete set of metric attributes. I think this test would be more explicit about the constraint, and would cover everything we cannot cover in code. Would this address your concern?

jyemin replied:
sgtm

jyemin reviewed (changes requested):
A few initial comments, but before I go further, I would like to see unit tests written for logic contained in the util (and sub) packages.  It seems easy to do, and would help catch any logic errors that might be missed in code review.

katcharov commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -207,51 +205,68 @@ public void start(final Map<String, String> props) {
     partitionMap = null;
     createPartitionMap(sourceConfig);
 
+    CommandListener statisticsCommandListener =
```
It seemed more appropriate to move part of the logic into `ResumeTokenUtils` as you suggested below, and the rest of the logic into a couple of methods on MongoSourceTask, similar to `commitRecord(`. Let me know if that doesn't address your concern. (Note that the remaining logic depends on `currentStatistics`, which is a non-final field of MongoSourceTask).

katcharov commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -207,51 +205,68 @@ public void start(final Map<String, String> props) {
+                          getTimestampFromResumeToken(resumeToken).asTimestamp().getTime();
+                      long opTime = response1.get("operationTime").asTimestamp().getTime();
+                      long offset = resumeTime - opTime;
+                      currentStatistics.getLatestOffsetSecs().sample(offset);
```
Good catch.

katcharov commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -207,51 +205,68 @@ public void start(final Map<String, String> props) {
+            // operationTime, clusterTime (no)
+            // changestreambatchtoken.resumecursor
+
+            Optional.ofNullable(response1)
```
Fair point. Refactored.

katcharov commented:
I added tests for the `MongoMBean`. These tests cover the behaviours of the individual metrics classes, and integration with the `MBeanServer`. I also added metrics checks, where appropriate, to the existing integration tests covering the source and sink tasks. In contrast to a mocked unit test of the tasks, this is less coupled to the implementation, and adds additional checks to those existing integration tests. I believe this overlap sufficiently covers the MBean functionality.

In those integration tests, I could not reliably get the counts to match in local testing. For example, the number of successful records was coming in under 100. There is a nice "eventually" mechanism (retries the checks) that I hoped to use for these metrics, but it was periodically failing even after substantially increasing the timeout to 50 seconds. I eventually weakened the tests by comparing against 0. On evergreen, there is also an error that is being reliably picked up in a locally-passing test that I do not expect to see an error in. I would consider these failures outside of the scope of the MBean functionality, though worth investigating.

jyemin commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -267,24 +249,32 @@ public void commandFailed(final CommandFailedEvent event) {
+    currentStatistics.getTaskInvocations().sample(taskTime.getElapsedTime(TimeUnit.MILLISECONDS));
+    lastTaskInvocation = Timer.start();
+
+    if (LOGGER.isDebugEnabled()) {
```
I think it would be better to only print the current statistic.  When we're in the copy state, stream statistics are empty.  And once we're in stream state, copy statistics no long change (I think).

katcharov replied:
Why would it be useful to separate copy and stream in JMX, but not in logging?

jyemin replied:
Separating is still useful.  I think the analogy would be if we were unregister the copy mbean once the copying phase is over.  And maybe that's a good idea too.

katcharov replied:
I changed it to log only the current.

I don't think we should unregister the copy mbean, because it might be sampled, and then accrue some counts, and then be unregistered before it is sampled again. (This is incidentally not an issue with logging, since no counts are accrued between logging and the switch from copy to stream.)

jyemin commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -828,9 +824,38 @@ BsonDocument getResumeToken(final MongoSourceConfig sourceConfig) {
+    }
+  }
+
+  private void mongoCommandSucceeded(final CommandSucceededEvent event) {
```
I still think that the entire CommandListener implementation should be broken out into a separate class and unit tested, but I can see the argument against, as it's tightly coupled to MongoSourceTask.

katcharov replied:
All stat invocations are now explicitly covered, including the getmore failure. I added stats checks to existing integration tests. The only one that wasn't already covered was `records-filtered` (a Kafka callback stat).

jyemin commented on src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java:
```diff
@@ -828,9 +824,38 @@ BsonDocument getResumeToken(final MongoSourceConfig sourceConfig) {
+    if (e instanceof MongoCommandException) {
+      if (doesNotSupportsStartAfter((MongoCommandException) e)) {
+        // silently ignore this expected exception, which is used to set this.supportsStartAfter
+        return;
```
It could be misleading to ignore these events entirely.  Did you consider whether to treat them as successful commands instead?

katcharov replied:
It seems that ideally, we might hit the server during initialization to see if it supports `startsAfter`. That wouldn't be as efficient. But if we did that, I would expect such a request to not be counted as a normal message, so on that basis I don't count it here. I agree that this might be confusing for users who enable server logging and are trying to match up message counts. But it seems more confusing to log this as a success when it would show up as an error in such logs. And counting it as an error when it is expected might needlessly worry users. Ignoring seems like the least-bad option, though I don't feel particularly strongly about this.

jyemin commented on src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java:
```diff
@@ -85,24 +86,44 @@ void stop() {
+            .mapToLong(ConnectRecord::timestamp)
+            .max();
+    if (latestRecord.isPresent()) {
+      long offsetMs = System.currentTimeMillis() - latestRecord.getAsLong();
```
I guess there is nothing we can do about the non-monotonicity of currentTimeMillis, if we want to track this stat?  Should we at least check for negative values here and ignore them?

katcharov replied:
Negative values seem fine, because if one clock is behind the other, we might start with a stable `-10000`, and perhaps set an alert if it moves up by 5000ms, signalling a lag spike. If we capped the negatives, any change would be obscured.

I imagine that in cases where non-monotonicity comes into play, such as a fast/slow clock being reset, we would want to see the reset reflected here.

jyemin replied:
I actually have no idea where the timestamp in ConnectRecord comes from.  Could it be from a different process or server?

katcharov replied:
The timestamp type associated with the timestamp is either the "create time" or the "log append" time. I don't know what it could be if not a reasonable approximation of the time the record was put into Kafka (and there's no other field we might use instead). [This SO answer and comment](https://stackoverflow.com/a/62936145) suggest that the default is "local epoch time of the producing system".

jyemin replied:
Yeah, then not a lot we can do here.  I'll resolve

katcharov commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MBeanServerUtils.java:
```diff
@@ -0,0 +1,120 @@
+    }
+  }
+
+  public static Map<String, Map<String, Long>> getMBeanAttributes(
```
This is used only by integration tests. If you prefer, I could move this code to `integrationTest`.

jyemin replied:
oh, yeah, if it's only used by tests let's move it.

katcharov replied:
I moved it, and then moved it back, since I needed it for a unit test in `test`, in addition to the usage in `integrationTest`. I think the integration test coverage is sufficient in this case.

rozza commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MBeanServerUtils.java:
```diff
@@ -0,0 +1,126 @@
+    // util class
+  }
+
+  public static String registerMBean(final DynamicMBean mBean, final String mBeanName) {
```
Nit: Annotate nullable?

rozza commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MBeanServerUtils.java:
```diff
@@ -0,0 +1,126 @@
+   * Thread names have the form "task-thread-MongoSinkConnector-0", where the number suffix is the
+   * id shown in Kafka's JMX MBean.
+   */
+  public static String taskIdFromCurrentThread() {
```
Nice work around the lack of official way to get the id.

rozza commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/AdditiveMetric.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+public class AdditiveMetric implements Metric {
```
I don't full grok how this works, can we add some docs to explain how it works and where / when it should be used.

katcharov replied:
Docs added. Perhaps the name `AdditiveMetric` isn't very clear, but I'm not sure what else to call this. The name metric itself might also be confusing - there are multiple values (MBean attributes) that arise out of a single measurement (`sample(long)`).

rozza reviewed (commented):
I think this is looking great - this is just a quick pass, looking at through the code. I'm going to debug / step through it before adding any more feedback as somethings I haven't fully grokked yet, which will come apparent when stepping through the code.

One question I have is do we have any idea how expensive adding this JMX reporting is?

katcharov commented on src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java:
```diff
@@ -85,24 +86,44 @@ void stop() {
+    statistics.getTaskInvocations().sample(taskTime.getElapsedTime(TimeUnit.MILLISECONDS));
+    lastTaskInvocation = Timer.start();
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(statistics.getName() + ": " + statistics.toJSON());
```
I wrote it this way because `toJSON` itself seemed relatively expensive, compared to the concatenation we'd be trying to save on here.

In response to your question about performance, I think it's negligible. We never increment counters in a tight loop. It's effectively an `AtomicLong` increment per stat. Usually this is once per poll, though iirc because of batching, retries, and the like, it might go up to "a few times" per poll.

katcharov replied:
Added a small comment.

rozza commented on src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MBeanServerUtils.java:
```diff
@@ -0,0 +1,133 @@
+    } catch (Exception e) {
+      // JMX might not be available
+      LOGGER.warn("Unable to register MBean " + mBeanName, e);
+      return null;
```
Is returning `null` handled / valid here? If I registerMBean and it returns null - will unregistering it cause a NPE instead of being a Noop?

jyemin replied:
unregisterMBean seems to handle unregistering a not-registered MBean.  It both checks for registration (line 83) and swallows exceptions (line 86)

rozza replied:
Ack - resolving.

katcharov replied:
The logger uses the mbean name, so this would log null in unusual circumstances. I returned an adjusted name.

rozza reviewed (changes requested):
Couple of small nits.

Once done LGTM

jyemin reviewed (approved):
LGTM
